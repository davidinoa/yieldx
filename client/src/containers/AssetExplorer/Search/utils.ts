interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
