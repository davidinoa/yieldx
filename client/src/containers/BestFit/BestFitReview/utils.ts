import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
