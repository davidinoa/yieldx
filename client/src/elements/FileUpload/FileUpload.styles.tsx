import styled, { keyframes } from "styled-components";

export const UploadOuter = styled.div`
  padding: 4rem;
  width: 100%;
  height: 100%;
`;

const fadeInKeyframe = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

export const UploadInner = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  cursor: pointer;
  grid-template-columns: max-content;
  text-align: right;
  align-items: center;
  justify-items: center;
  justify-content: center;

  padding: 4rem;
  border: 1px dashed transparent;
  animation: 1s ${fadeInKeyframe} ease-in-out;

  box-sizing: border-box;
  border-radius: 10px;

  -webkit-transition: color 0.25s;
  transition: color 0.25s;

  background: none;

  box-sizing: border-box;

  position: relative;
  &::before,
  &::after {
    box-sizing: inherit;
    content: "";
    position: absolute;
    width: 100%;
    height: 100%;

    border: 1px dashed transparent;
    border-radius: 10px;
    width: 0;
    height: 0;
  }
  &::before {
    border-radius: 10px;
    top: 0;
    left: 0;
  }
  &::after {
    border-radius: 10px;
    bottom: 0;
    right: 0;
  }
  &:hover::before,
  &:hover::after {
    width: 100%;
    height: 100%;
    border-style: dashed;
  }
  &:hover::before {
    border-top-color: var(--primary-hover);
    border-right-color: var(--primary-hover);
    -webkit-transition: width 0.125s ease-out, height 0.125s ease-out 0.125s;
    transition: width 0.125s ease-out, height 0.125s ease-out 0.125s;
  }
  &:hover::after {
    border-bottom-color: var(--primary-hover);
    border-left-color: var(--primary-hover);
    -webkit-transition: border-color 0s ease-out 0.25s,
      width 0.125s ease-out 0.25s, height 0.125s ease-out 0.375s;
    transition: border-color 0s ease-out 0.25s, width 0.125s ease-out 0.25s,
      height 0.125s ease-out 0.375s;
  }
`;

export const SubTitle = styled.h3`
  font-size: 13px;
  line-height: 13px;
  /* identical to box height */

  text-align: center;
  letter-spacing: 0.05em;

  /* Black / Light-4 */
  color: #8290b0;
`;

export const Note = styled.p`
  margin-top: 0.5rem;
  font-size: 10px;
  line-height: 10px;
  /* identical to box height */

  /* Black / Light-4 */

  color: #8290b0;
`;
