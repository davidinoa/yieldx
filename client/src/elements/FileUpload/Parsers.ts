import { Sheet2JSONOpts } from "xlsx/types";

export enum SupportedParser {
  XLSX = "xlsx",
  JSON = "json",
  TEXT = "text",
  PDF = "pdf",
  RAW = "binary"
}

export type ParserMap = { [key: string]: SupportedParser };
export const TextMimeType: ParserMap = {
  "text/plain": SupportedParser.XLSX,
  "text/csv": SupportedParser.XLSX,
  "application/json": SupportedParser.JSON,
  "application/ld+json": SupportedParser.JSON,
  "text/xml": SupportedParser.XLSX,
  "text/html": SupportedParser.RAW,
  "text/css": SupportedParser.RAW,
  "application/xml": SupportedParser.XLSX
};

export const BinaryMimeType: ParserMap = {
  "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet":
    SupportedParser.XLSX,
  "application/vnd.ms-excel": SupportedParser.XLSX,
  "application/pdf": SupportedParser.XLSX
};

export const ExtensionHandlers: ParserMap = {
  "*": SupportedParser.RAW,
  ".pdf": SupportedParser.XLSX,
  ".json": SupportedParser.JSON,
  ".jsonld": SupportedParser.JSON,
  ".xlsx": SupportedParser.XLSX,
  ".xlsb": SupportedParser.XLSX,
  ".xlsm": SupportedParser.XLSX,
  ".xls": SupportedParser.XLSX,
  ".xml": SupportedParser.XLSX,
  ".csv": SupportedParser.XLSX,
  ".txt": SupportedParser.XLSX,
  ".ods": SupportedParser.XLSX,
  ".fods": SupportedParser.XLSX,
  ".uos": SupportedParser.XLSX,
  ".sylk": SupportedParser.XLSX,
  ".dif": SupportedParser.XLSX,
  ".dbf": SupportedParser.XLSX,
  ".prn": SupportedParser.XLSX,
  ".qpw": SupportedParser.XLSX,
  ".123": SupportedParser.XLSX,
  ".wb*": SupportedParser.XLSX,
  ".wq*": SupportedParser.XLSX,
  ".html": SupportedParser.XLSX,
  ".htm": SupportedParser.XLSX
};
export type FileMetadata = {
  type: MimeTypes;
  name: string;
  size: number;
  parser: SupportedParser | Parser;
};

export type ParseResult = {
  meta: FileMetadata;
  data: any;
  [key: string]: any;
};

export interface Parser {
  (data: string | ArrayBuffer | null, meta: FileMetadata): Promise<ParseResult>;
}

export type MimeTypes = Exclude<keyof typeof MimeHandlers, number>;
export type FileTypes = Exclude<keyof typeof ExtensionHandlers, number>;
export const SupportedExtensionRegex = new RegExp(
  "[.](pdf|json|jsonld|xlsx|xlsb|xlsm|xls|xml|csv|txt|ods|fods|uos|sylk|dif|dbf|prn|qpw|123|wb[*]|wq[*]|html|htm)$"
);

export function formatExtension(ext: string): FileTypes {
  return `.${ext}` as FileTypes;
}

export function getParserByFileName(name: string): SupportedParser {
  if (SupportedExtensionRegex.test(name)) {
    const matches = SupportedExtensionRegex.exec(name);
    if (matches?.length) return ExtensionHandlers[formatExtension(matches[0])];
  }
  return SupportedParser.RAW;
}

export function getParserForFile(
  file: File,
  parser?: SupportedParser | Parser
) {
  // Custom parser
  if (parser && parser instanceof Function) return parser;
  if ((file.type as MimeTypes) in MimeHandlers)
    return MimeHandlers[file.type as MimeTypes];
  return getParserByFileName(file.name);
}

export const MimeHandlers = { ...TextMimeType, ...BinaryMimeType };
export const SupportedMimeTypes = Object.keys(MimeHandlers);

export const DefaultMetadata = Object.freeze({
  size: 0,
  type: "text/plain",
  name: "",
  parser: SupportedParser.XLSX
});

export function ParseXlsx(
  data: string | ArrayBuffer | null,
  dataType: "binary" | "buffer" | "base64" | "string" | "file" = "binary",
  headers?: string[]
) {
  return import("xlsx").then(imp => {
    const XLSX = imp.default;
    const wb = XLSX.read(data, { type: dataType });
    /* Get first worksheet */
    const wsname = wb.SheetNames[0];
    const ws = wb.Sheets[wsname];
    /* Convert array of arrays */
    const options: Sheet2JSONOpts = { blankrows: false };
    if (headers?.length) options.header = headers;
    return XLSX.utils.sheet_to_json(ws, options);
    /* Update state */
  });
}
