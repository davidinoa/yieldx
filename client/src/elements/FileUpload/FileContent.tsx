import React, { useState } from "react";
import { Table } from "elements/Table/Table";
import { ParseResult } from "elements/FileUpload/Parsers";
import styled from "styled-components";
import { Collapse } from "react-bootstrap";
import clsx from "clsx";
import { ChevronDown, ChevronUp } from "elements/Icon";
import { LabeledField } from "elements/Field/Labeled";

const ContentWrapper = styled.div`
  min-width: 300px,
  margin-top: 2rem
  border: 1px solid --var(--border-2);
`;
const FileHeader = styled.div`
  margin: 0 2rem;
  text-align: left;
  margin-bottom: 1rem;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
`;

const FilePreview = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: stretch;
  > * {
    width: 100%;
  }
  nav {
    margin-top: 0;
    margin-bottom: 1rem;
  }
  border: 1px solid rgba(120, 120, 120, 0.6);
  overflow: scroll;
`;
const CollapseToggle = styled.div`
  margin: 1rem;
  cursor: pointer;
  &::hover {
    color: var(--hover);
  }
`;

export function FileContent({ data, meta }: ParseResult) {
  const [open, setOpen] = useState(false);

  return (
    <ContentWrapper className="justify-content-center">
      <CollapseToggle onClick={() => setOpen(!open)}>
        <FileHeader>
          <LabeledField
            label="File"
            reverse
            value={`${meta.name} (${data.length} rows)`}
          />

          {open ? <ChevronUp /> : <ChevronDown />}
        </FileHeader>
      </CollapseToggle>
      <Collapse in={open}>
        <div className={clsx("m-0 p-0")}>
          <Table data={data} />
        </div>
      </Collapse>
    </ContentWrapper>
  );
}

export function FileList({ files }: { files: ParseResult[] }) {
  return (
    <FilePreview>
      {files.reverse().map((file: any) => (
        <FileContent key={file.meta.name} {...file} />
      ))}
    </FilePreview>
  );
}
