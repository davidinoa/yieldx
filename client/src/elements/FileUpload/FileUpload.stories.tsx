import React, { useState } from "react";
import { Story, Meta } from "@storybook/react/types-6-0";
import { FileUpload, FileUploadProps } from "elements/FileUpload/FileUpload";
import { ParseResult } from "elements/FileUpload/Parsers";
import { FileList } from "./FileContent";

const exampleStyle = {
  margin: "80px",
  border: "1px solid #cccccc55",
  borderRadius: 10
};

const Template: Story<FileUploadProps> = args => {
  const [results, setResults] = useState<ParseResult[]>([]);
  const { onFileComplete: _ignore, ...props } = args;
  return (
    <div style={exampleStyle}>
      <FileUpload
        {...props}
        onFileComplete={res => {
          setResults([...results, res]);
          return undefined;
        }}
      />
      {results.length > 0 ? <FileList files={results} /> : null}
    </div>
  );
};

export const Primary = Template.bind({});
Primary.args = {};
export default {
  component: FileUpload,
  title: "Elements/FileUpload"
} as Meta;
