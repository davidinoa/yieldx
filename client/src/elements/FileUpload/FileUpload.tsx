import React, { useCallback, useState } from "react";
import { Row, Button } from "react-bootstrap";
import { useDropzone } from "react-dropzone";
import { ReactComponent as UploadIcon } from "assets/svg/upload.svg";
import clsx from "clsx";
import {
  DefaultMetadata,
  FileMetadata,
  FileTypes,
  getParserForFile,
  MimeTypes,
  Parser,
  ParseResult,
  ParseXlsx,
  SupportedMimeTypes,
  SupportedParser,
  TextMimeType
} from "elements/FileUpload/Parsers";
import * as Styles from "./FileUpload.styles";

export type FileUploadProps = {
  onFileComplete: (parsed: ParseResult) => void;
  parser?: SupportedParser | Parser;
  maxSize?: number;
  acceptedTypes?:
    | (MimeTypes | FileTypes | string)
    | (MimeTypes | FileTypes | string)[];
  headers?: string[];
};

export function FileUpload({
  onFileComplete,
  parser,
  headers,
  maxSize = 2e6,
  acceptedTypes = SupportedMimeTypes
}: FileUploadProps) {
  const [errors, setError] = useState<string[]>([]);
  const [results, setResults] = useState<ParseResult[]>([]);

  const onError = useCallback(
    err => {
      let newErrors: string[] = [];
      if (err instanceof Array) {
        newErrors = err.map((e: { errors: any }) =>
          e.errors.map(({ message }: { message: string }) => message)
        );
      } else {
        err instanceof Error
          ? newErrors.push(err.message)
          : newErrors.push(err);
      }
      setError([...errors, ...newErrors]);
    },
    [setError, errors]
  );
  const callbackWithMessage = useCallback(
    (message: string) => () => onError(message),
    [onError]
  );
  const onLoad = useCallback(
    (
      r: FileReader,
      meta: FileMetadata = DefaultMetadata
    ): Promise<ParseResult> => {
      if (!r.result) {
        return Promise.reject(new Error("File appears to be empty."));
      }
      let parsedResult: Promise<ParseResult>;
      if (meta.parser instanceof Function) {
        return meta.parser(r.result, meta);
      }
      switch (meta.parser) {
        case SupportedParser.XLSX:
          parsedResult = ParseXlsx(r.result, "binary", headers).then(data => {
            return { meta, data } as ParseResult;
          });
          break;
        case SupportedParser.JSON:
          parsedResult = new Promise((resolve, reject) => {
            try {
              resolve({
                meta,
                data: JSON.parse(r.result as string)
              } as ParseResult);
            } catch (error) {
              reject(error);
            }
          });
          break;
        case SupportedParser.RAW:
        default:
          parsedResult = Promise.resolve({
            meta,
            data: r.result
          } as ParseResult);
          break;
      }
      return parsedResult;
    },
    [headers]
  );
  const onParseComplete = useCallback(
    parsed => {
      onFileComplete(parsed);
      setResults([...results, parsed]);
    },
    [setResults, results, onFileComplete]
  );

  const onDrop = useCallback(
    (acceptedFiles: File[]) => {
      acceptedFiles.forEach(file => {
        const metadata = {
          name: file.name,
          type: file.type as MimeTypes,
          size: file.size,
          parser: getParserForFile(file, parser)
        };
        const reader = new FileReader();
        reader.onload = () =>
          onLoad(reader, metadata)
            .then(onParseComplete)
            .catch(err => setError([...errors, err.message]));
        reader.onabort = callbackWithMessage("file reading was aborted");
        reader.onerror = callbackWithMessage("failed to read the file.");

        metadata.type in TextMimeType
          ? reader.readAsText(file)
          : reader.readAsBinaryString(file);
      });
    },
    [parser, errors, onLoad, onParseComplete, callbackWithMessage]
  );

  const { getRootProps, getInputProps, open } = useDropzone({
    maxSize,
    accept: acceptedTypes || "*",
    onDrop,
    onDropRejected: onError
  });
  return (
    <span className={clsx("draw")}>
      <Styles.UploadInner
        data-testid="FileUpload"
        {...getRootProps({ refKey: "innerRef" })}
      >
        <Row>
          <UploadIcon />
        </Row>
        <Row className={clsx("mt-2")}>
          <h2 style={{ fontWeight: "inherit" }}>Drag & drop you file here</h2>
        </Row>
        <Row>
          <Styles.SubTitle>Drag & drop you file here</Styles.SubTitle>
        </Row>
        <Row className={clsx("mt-3 mb-2")}>
          <h3>OR</h3>
        </Row>
        <Row>
          <Button onClick={open}>Browse files</Button>
        </Row>
        <Row>
          <Styles.Note>
            Maximum file size is {Math.round(maxSize / 1e6)}mb
          </Styles.Note>
        </Row>
        <input {...getInputProps()} />
        {errors.map(e => {
          return <Row key={e}>Error: {e}</Row>;
        })}
      </Styles.UploadInner>
    </span>
  );
}
