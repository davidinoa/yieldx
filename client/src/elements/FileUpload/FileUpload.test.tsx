import React from "react";
import userEvent from "@testing-library/user-event";
import Dropzone from "react-dropzone";
import { act, render, waitFor } from "test/utils";
import { fireEvent } from "@testing-library/react";
import { FileUpload } from "elements/FileUpload/FileUpload";
import { SupportedParser } from "elements/FileUpload/Parsers";

test("snapshot test", async () => {
  const onFileComplete = jest.fn();

  const ui = (
    <FileUpload
      onFileComplete={parsed => {
        onFileComplete(parsed);
        console.error(parsed);
      }}
    />
  );
  const { asFragment, container, rerender } = render(ui);
  expect(asFragment()).toMatchSnapshot();
});
test.skip("invoke onDragEnter when dragenter event occurs", async () => {
  const onFileComplete = jest.fn();

  const ui = (
    <FileUpload
      parser={(data, meta) => {
        // throw new Error("yeahh");
        return new Promise(
          res =>
            onFileComplete() &&
            res({
              data: [1],
              meta: {
                name: "this.jsx",
                type: "application/json",
                size: 1048,
                parser: SupportedParser.JSON
              }
            })
        );
      }}
      onFileComplete={parsed => {
        onFileComplete(parsed);
        console.error(parsed);
      }}
    />
  );
  const { rerender, findByTestId } = render(ui);

  const file = new File([JSON.stringify({ ping: true })], "ping.json", {
    type: "application/json"
  });
  const data = mockData([file]);
  const dropzone = await findByTestId("FileUpload");
  if (dropzone) dispatchEvt(dropzone, "dragenter", data);
  await flushPromises(rerender, ui);

  // expect(dropzone).toEqual({});
  expect(onFileComplete).toHaveBeenCalled();
});

async function flushPromises<T extends any>(
  rerender: (param: T) => void,
  ui: T
) {
  await act(() => waitFor(() => rerender(ui)));
}

function dispatchEvt(
  node: Element | Node | Document | Window,
  type: string,
  data: any
) {
  const event = new Event(type, { bubbles: true });
  Object.assign(event, data);
  fireEvent(node, event);
}

function mockData(files: any[]) {
  return {
    dataTransfer: {
      files,
      items: files.map((file: any) => ({
        kind: "file",
        type: file.type,
        getAsFile: () => file
      })),
      types: ["Files"]
    }
  };
}
