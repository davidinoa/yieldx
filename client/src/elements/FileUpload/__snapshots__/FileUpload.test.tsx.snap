// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invoke onDragEnter when dragenter event occurs 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
  grid-template-columns: max-content;
  text-align: right;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4rem;
  border: 1px dashed transparent;
  -webkit-animation: 1s cJoqxJ ease-in-out;
  animation: 1s cJoqxJ ease-in-out;
  box-sizing: border-box;
  border-radius: 10px;
  -webkit-transition: color 0.25s;
  -webkit-transition: color 0.25s;
  transition: color 0.25s;
  background: none;
  box-sizing: border-box;
  position: relative;
}

.c0::before,
.c0::after {
  box-sizing: inherit;
  content: "";
  position: absolute;
  width: 100%;
  height: 100%;
  border: 1px dashed transparent;
  border-radius: 10px;
  width: 0;
  height: 0;
}

.c0::before {
  border-radius: 10px;
  top: 0;
  left: 0;
}

.c0::after {
  border-radius: 10px;
  bottom: 0;
  right: 0;
}

.c0:hover::before,
.c0:hover::after {
  width: 100%;
  height: 100%;
  border-style: dashed;
}

.c0:hover::before {
  border-top-color: var(--primary-hover);
  border-right-color: var(--primary-hover);
  -webkit-transition: width 0.125s ease-out,height 0.125s ease-out 0.125s;
  -webkit-transition: width 0.125s ease-out,height 0.125s ease-out 0.125s;
  transition: width 0.125s ease-out,height 0.125s ease-out 0.125s;
}

.c0:hover::after {
  border-bottom-color: var(--primary-hover);
  border-left-color: var(--primary-hover);
  -webkit-transition: border-color 0s ease-out 0.25s, width 0.125s ease-out 0.25s,height 0.125s ease-out 0.375s;
  -webkit-transition: border-color 0s ease-out 0.25s,width 0.125s ease-out 0.25s, height 0.125s ease-out 0.375s;
  transition: border-color 0s ease-out 0.25s,width 0.125s ease-out 0.25s, height 0.125s ease-out 0.375s;
}

.c1 {
  font-size: 13px;
  line-height: 13px;
  text-align: center;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  color: #8290b0;
}

.c2 {
  margin-top: 0.5rem;
  font-size: 10px;
  line-height: 10px;
  color: #8290b0;
}

<span
    class="draw"
    data-testid="FileUpload"
  >
    <div
      class="c0"
      tabindex="0"
    >
      <div
        class="row"
      >
        <svg>
          upload.svg
        </svg>
      </div>
      <div
        class="mt-2 row"
      >
        <h2
          style="font-weight: inherit;"
        >
          Drag & drop you file here
        </h2>
      </div>
      <div
        class="row"
      >
        <h3
          class="c1"
        >
          Drag & drop you file here
        </h3>
      </div>
      <div
        class="mt-3 mb-2 row"
      >
        <h3>
          OR
        </h3>
      </div>
      <div
        class="row"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Browse files
        </button>
      </div>
      <div
        class="row"
      >
        <p
          class="c2"
        >
          Maximum file size is 2mb
        </p>
      </div>
      <input
        accept="text/plain,text/csv,application/json,application/ld+json,text/xml,text/html,text/css,application/xml,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,application/vnd.ms-excel,application/pdf"
        autocomplete="off"
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
    </div>
  </span>
</DocumentFragment>
`;

exports[`snapshot test 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
  grid-template-columns: max-content;
  text-align: right;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4rem;
  border: 1px dashed transparent;
  -webkit-animation: 1s cJoqxJ ease-in-out;
  animation: 1s cJoqxJ ease-in-out;
  box-sizing: border-box;
  border-radius: 10px;
  -webkit-transition: color 0.25s;
  -webkit-transition: color 0.25s;
  transition: color 0.25s;
  background: none;
  box-sizing: border-box;
  position: relative;
}

.c0::before,
.c0::after {
  box-sizing: inherit;
  content: "";
  position: absolute;
  width: 100%;
  height: 100%;
  border: 1px dashed transparent;
  border-radius: 10px;
  width: 0;
  height: 0;
}

.c0::before {
  border-radius: 10px;
  top: 0;
  left: 0;
}

.c0::after {
  border-radius: 10px;
  bottom: 0;
  right: 0;
}

.c0:hover::before,
.c0:hover::after {
  width: 100%;
  height: 100%;
  border-style: dashed;
}

.c0:hover::before {
  border-top-color: var(--primary-hover);
  border-right-color: var(--primary-hover);
  -webkit-transition: width 0.125s ease-out,height 0.125s ease-out 0.125s;
  -webkit-transition: width 0.125s ease-out,height 0.125s ease-out 0.125s;
  transition: width 0.125s ease-out,height 0.125s ease-out 0.125s;
}

.c0:hover::after {
  border-bottom-color: var(--primary-hover);
  border-left-color: var(--primary-hover);
  -webkit-transition: border-color 0s ease-out 0.25s, width 0.125s ease-out 0.25s,height 0.125s ease-out 0.375s;
  -webkit-transition: border-color 0s ease-out 0.25s,width 0.125s ease-out 0.25s, height 0.125s ease-out 0.375s;
  transition: border-color 0s ease-out 0.25s,width 0.125s ease-out 0.25s, height 0.125s ease-out 0.375s;
}

.c1 {
  font-size: 13px;
  line-height: 13px;
  text-align: center;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  color: #8290b0;
}

.c2 {
  margin-top: 0.5rem;
  font-size: 10px;
  line-height: 10px;
  color: #8290b0;
}

<span
    class="draw"
  >
    <div
      class="c0"
      data-testid="FileUpload"
      tabindex="0"
    >
      <div
        class="row"
      >
        <svg>
          upload.svg
        </svg>
      </div>
      <div
        class="mt-2 row"
      >
        <h2
          style="font-weight: inherit;"
        >
          Drag & drop you file here
        </h2>
      </div>
      <div
        class="row"
      >
        <h3
          class="c1"
        >
          Drag & drop you file here
        </h3>
      </div>
      <div
        class="mt-3 mb-2 row"
      >
        <h3>
          OR
        </h3>
      </div>
      <div
        class="row"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Browse files
        </button>
      </div>
      <div
        class="row"
      >
        <p
          class="c2"
        >
          Maximum file size is 2mb
        </p>
      </div>
      <input
        accept="text/plain,text/csv,application/json,application/ld+json,text/xml,text/html,text/css,application/xml,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,application/vnd.ms-excel,application/pdf"
        autocomplete="off"
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
    </div>
  </span>
</DocumentFragment>
`;
